rust/src/delta.rs

```diff
@@ -1123,6 +1059,26 @@ impl DeltaTable {
             .get_configuration())
     }
 
+    /// Vacuum the delta table. See [`VacuumBuilder`] for more information.
+    pub async fn vacuum(
+        &mut self,
+        retention_hours: Option<u64>,
+        dry_run: bool,
+        enforce_retention_duration: bool,
+    ) -> Result<Vec<String>, DeltaTableError> {
+        let state = std::mem::take(&mut self.state);
+        let mut plan = VacuumBuilder::new(self.object_store(), state)
+            .with_dry_run(dry_run)
+            .with_enforce_retention_duration(enforce_retention_duration);
+        if let Some(hours) = retention_hours {
+            plan = plan.with_retention_period(Duration::hours(hours as i64));
+        }
+
+        let (table, metrics) = plan.await?;
+        self.state = table.state;
```

Comment:
Here I am a little bit concernbed about the error handling. If we fail duing the execution, we are still left with a DeltaTable with an empty state. If users wanted to recover from a failure and keep using the table, they would probably see unexpexted results.

Once we manage the state in arrow, we can likely have near zero-cost clones of the state and mitigate the issue. For one a way to handle this could be to handle the error internally, and load the previous table state again. But not sure what users wold expect / want to see.

The same applies on the python side.

Reply:
That would make the mutation operations like `merge()` and `process_action()` less efficient.

Reply:
Can we clone the state for now, and as a follow up let's figure out how to do this more efficiently?